### ISSUE TITLE
Add support for matrix square root

### ISSUE THREAD

**Issue by emchristiansen** [tensorflow/tensorflow#9202]
Please consider adding a [matrix square root](https://en.wikipedia.org/wiki/Square_root_of_a_matrix) operation, with gradients.
It will make it possible to implement [stable whitening](https://arxiv.org/abs/1512.00809), which could be broadly useful, in addition to being useful for my particular problem :)
Note, Cholesky whitening is currently supported in TensorFlow, but I'm not aware of any guarantees it provides regarding the correspondences between whitened and non-whitened data.

It appears Eigen [already has a matrix square root function](https://eigen.tuxfamily.org/dox/unsupported/group__MatrixFunctions__Module.html#matrixbase_sqrt), so this might not be too hard to implement.

**Comment by asimshankar** [comment]
CC @rmlarsen @agarwal-ashish in case they have any comments here.

Generally sounds like a good idea, contributions welcome.

**Comment by bhairavmehta95** [comment]
Has this been claimed? I am looking to make my first contribution and I think this would be a good one to get started with.

**Comment by bhairavmehta95** [comment]
@rmlarsen @agarwal-ashish @asimshankar 

I was hoping to essentially redo the Eigen square root function. But, I had a quick question as a newer contributor:
Should the PR only be made when the function is fully supported in all C, C++, and Python?

**Comment by theflofly** [comment]
@bhairavmehta95 same question here.

I'd say that you can provide the op in c++ for instance and an other contributor will add the python part...

**Comment by bhairavmehta95** [comment]
@theflofly thanks for the response. I guess I will assign myself informally to the Python implementation, and move on from there!

**Comment by yaroslavvb** [comment]
BTW, for symsqrt I've been using function below. No gradients, but will be after SVD gradients are supported as in https://github.com/tensorflow/tensorflow/issues/6503

```
def symsqrt(mat, eps=1e-7):
  """Symmetric square root."""
  s, u, v = tf.svd(mat)
  # sqrt is unstable around 0, just use 0 in such case
  si = tf.where(tf.less(s, eps), s, tf.sqrt(s))
  return u @ tf.diag(si) @ tf.transpose(v)
```
If you just want to whiten data, and backprop through it, you can use Cholesky.
Whitening is accomplished by using modified Cholesky Decomposition which is just a couple of arithmetic ops on top of Cholesky, here's a [mini tutorial](https://wolfr.am/lbQ9mbnl)

**Comment by emchristiansen** [comment]
@yaroslavvb thanks for the mini tutorial; it looks like it should be more efficient than just computing `cholesky(inverse(covariance_matrix))`.

However, for a dataset with dimension > 1 there are infinitely many whitening matrices, because a rotation of white data is still white.
The Cholesky approach produces just one of these possible matrices.
As mentioned in the first post, a matrix square root operation would make it possible to produce matrices with more desirable properties, e.g. a matrix which maximizes correlation between whitened and non-whitened data.

**Comment by yaroslavvb** [comment]
@emchristiansen The`symsqrt` function above will produce whitening matrix of maximum trace

**Comment by yaroslavvb** [comment]
BTW, this [paper](https://arxiv.org/pdf/1512.00809.pdf) talks about constraints enforced by Cholesky whitening which are indeed different from square root whitening. The `symsqrt` version above should be equivalent to matrix square root when the matrix is invertible (it's "ZCA whitening" in the paper)

**Comment by emchristiansen** [comment]
@bhairavmehta95 do you intend to implement this?
If not I might give it a shot or try another approach.

@yaroslavvb that's the paper I linked to in the first post :)

**Comment by lakshayg** [comment]
Is someone currently working on this issue at the moment @bhairavmehta95 @emchristiansen? I intent to start contributing to tensorflow. I have tried to figure out the roadmap to implement the matrix_sqrt feature and this is what I have come up with:

- [ ] Create `matrix_sqrt_op.cc` in `tensorflow/tensorflow/core/kernels`
- [ ] Add `matrix_sqrt_op` entry to `tensorflow/tensorflow/core/kernels/BUILD`
- [ ] Write a test function in `tensorflow/tensorflow/core/ops/linalg_ops_test.cc`
- [ ] Register op in `tensorflow/tensorflow/core/ops/linalg_ops.cc`
- [ ] Add a `MatrixSqrt` entry to `tensorflow/tensorflow/core/ops/ops.pbtxt`
- [ ] Define `_MatrixSqrtGrad(op, grad)` in `tensorflow/tensorflow/python/ops/linalg_grad.py`
- [ ] Write a test function in `tensorflow/python/kernel_tests/linalg_grad_test.py`
- [ ] Write a go wrapper in `tensorflow/go/op/wrappers.go`

Please let me know if what I have figured out is correct or if there are any changes that should be made to this plan.

**Comment by emchristiansen** [comment]
@lakshayg I'm not currently working on this, and I'd guess @bhairavmehta95 isn't working on it either. So I think it's yours if you want it :)

I'm not familiar enough with the C++ side of TensorFlow to meaningfully comment on your plan, but it seems like a good start at least.
If I were you (and I'm guessing this is what you did), I'd just model the code on another unary matrix op, e.g. matrix inverse.

**Comment by jonathanwyatt16** [comment]
It looks like over a year has passed since the last comment so I went ahead and added matrix square root op support in PR #20611. Any feedback would be much appreciated!

**Comment by wt-huang** [comment]
Closing as this is resolved
